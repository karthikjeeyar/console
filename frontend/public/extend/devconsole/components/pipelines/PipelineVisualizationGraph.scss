$border-color: var(--pf-global--BorderColor--light);
$gutter: 1.8em;

.odc-pipeline__graph {
  background: white;
  padding: 10px;
  margin: 10px 0;
  width: 80vw;
  height: 20em;
  overflow: auto;
  position: relative;
  -webkit-overflow-scrolling: touch;

  // reset
  & ul {
    list-style: none;
    padding: 0;
    margin: 0;
  }

  & .odc-pipeline__stages {
    display: inline-flex;
  }

  & .odc-pipeline__stage,
  .odc-pipeline__stage--is-parallel {
    margin: 0 2em;

    &:not(:first-child) {
      & .odc-pipeline__task:first-child {
        &::before {
          content: '';
          position: absolute;
          top: 1em;

          border-top: 1px solid $border-color;
          width: $gutter * 1.7 + 1;
          height: 0;
        }
      }
    }
  }

  .odc-pipeline__stage--is-parallel {
    & .odc-pipeline__task {
      &:first-child {
        &::before,
        &::after {
          content: '';
          position: absolute;
          top: 1em;
          border-top: 1px solid $border-color;
          width: $gutter / 2 + 1;
          height: 0;
        }
        &::before {
          left: 0;
          transform: translateX(-100%);
        }
        &::after {
          right: 0;
          transform: translateX(100%);
        }
      }
    }
    &:last-child {
      & .odc-pipeline__task:first-child {
        &::after {
          content: '';
          width: $gutter / 2 + 0.05;
        }
      }
    }
  }
  &:not(.odc-pipeline__stage--is-parallel) {
    & .odc-pipeline__task:first-child {
      &::before {
        left: 0;
        transform: translateX(-100%);
      }
    }
  }
}
