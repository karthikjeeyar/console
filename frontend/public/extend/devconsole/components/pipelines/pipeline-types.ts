/* eslint-disable no-unused-vars, no-undef */
import { K8sResourceKind } from '../../../../module/k8s';

export interface PipelineVisualizationProps {
  pipeline?: K8sResourceKind;
}

export interface PipelineVisualizationTaskProp {
  name: string;
  resources?: {
    inputs?: object;
    outputs?: object;
  };
  params?: object;
  runAfter?: string[];
  taskRef: {
    name: string;
  };
}
