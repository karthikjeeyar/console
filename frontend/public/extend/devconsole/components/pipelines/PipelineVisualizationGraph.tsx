/* eslint-disable no-unused-vars, no-undef */
import * as React from 'react';
import { PipelineVisualizationTask } from './PipelineVisualizationTask';
import './PipelineVisualizationGraph.scss';
import { Firehose } from '../../../../components/utils';

interface Resources {
  inputs?: Resource[];
  outputs?: Resource[];
}

interface Resource {
  name: string;
  resource?: string;
  from?: string[];
}
export interface PipelineVisualizationTaskItem {
  name: string;
  resources?: Resources;
  params?: object;
  runAfter?: string[];
  taskRef: {
    name: string;
  };
}
export interface PipelineVisualizationGraphProps {
  graph: Array<PipelineVisualizationTaskItem[]>;
  namespace: string;
}

export const PipelineVisualizationGraph: React.FC<PipelineVisualizationGraphProps> = ({
  graph,
  namespace,
}) => {
  return (
    <div className="odc-pipeline__graph">
      <div className="odc-pipeline__stages">
        <div className="odc-pipeline__stage">
          <div className="odc-pipeline__task input-node" />
        </div>
        {graph.map((stage, i) => {
          const stageClassName = 'odc-pipeline__stage';
          return (
            <div
              className={stage.length <= 1 ? stageClassName : `${stageClassName}--is-parallel`}
              key={`stage:${i}`}
            >
              <ul className="odc-pipeline__stage-column">
                {stage.map((task, j) => {
                  return (
                    <Firehose
                      key={`${task.taskRef.name}:${i + j}`}
                      resources={[
                        {
                          kind: 'Task',
                          name: task.taskRef.name,
                          namespace,
                          prop: 'task',
                        },
                      ]}
                    >
                      <PipelineVisualizationTask task={task.taskRef.name} namespace={namespace} />
                    </Firehose>
                  );
                })}
              </ul>
            </div>
          );
        })}
      </div>
    </div>
  );
};
