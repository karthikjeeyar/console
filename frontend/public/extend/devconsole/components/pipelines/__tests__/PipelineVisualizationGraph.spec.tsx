/* eslint-disable no-unused-vars, no-undef */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import * as Renderer from 'react-test-renderer';
import PipelineVisualizationGraph from '../PipelineVisualizationGraph';
import { PipelineVisualizationProps } from '../pipeline-types';
import { mockPipelineGraph } from '../__mocks__/PipelineMocks';

jest.mock('react-dom', () => ({
  findDOMNode: () => ({}),
  createPortal: (node) => node,
}));

describe('PipelineVisualizationGraph', () => {
  const props = {
    namespace: 'test',
    graph: mockPipelineGraph,
  };
  let wrapper: ShallowWrapper<PipelineVisualizationProps>;
  beforeEach(() => {
    wrapper = shallow(<PipelineVisualizationGraph {...props} />);
  });

  it('renders a Pipeline visualization graph', () => {
    expect(wrapper.exists()).toBeTruthy();
  });

  it('should contain right number of stages', () => {
    const noOfStages = wrapper.find('.odc-pipeline-stage').length - 1; // to exclude first static input node
    expect(noOfStages).toEqual(mockPipelineGraph.length);
  });

  it('should match the previous snapshot', () => {
    const tree = Renderer.create(<PipelineVisualizationGraph {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
