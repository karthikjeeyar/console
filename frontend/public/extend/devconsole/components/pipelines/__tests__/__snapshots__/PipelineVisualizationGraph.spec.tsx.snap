// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelineVisualizationGraph should match the previous snapshot 1`] = `
<div
  className="odc-pipeline__graph"
>
  <div
    className="odc-pipeline__stages"
  >
    <div
      className="odc-pipeline__stage"
    >
      <div
        className="odc-pipeline__task input-node"
      />
    </div>
    <div
      className="odc-pipeline__stage"
    >
      <ul
        className="odc-pipeline__stage-column"
      >
        <li
          className="odc-pipeline__task"
        >
          <div
            className="title"
          >
            <span
              className="truncate"
            >
              start-app
            </span>
            <div
              className="pf-c-tooltip"
              role="tooltip"
            >
              <div
                className="pf-c-tooltip__arrow"
              />
              <div
                className="pf-c-tooltip__content"
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="odc-pipeline__stage--is-parallel"
    >
      <ul
        className="odc-pipeline__stage-column"
      >
        <li
          className="odc-pipeline__task"
        >
          <div
            className="title"
          >
            <span
              className="truncate"
            >
              test-app-1
            </span>
            <div
              className="pf-c-tooltip"
              role="tooltip"
            >
              <div
                className="pf-c-tooltip__arrow"
              />
              <div
                className="pf-c-tooltip__content"
              />
            </div>
          </div>
        </li>
        <li
          className="odc-pipeline__task"
        >
          <div
            className="title"
          >
            <span
              className="truncate"
            >
              test-app-2
            </span>
            <div
              className="pf-c-tooltip"
              role="tooltip"
            >
              <div
                className="pf-c-tooltip__arrow"
              />
              <div
                className="pf-c-tooltip__content"
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="odc-pipeline__stage--is-parallel"
    >
      <ul
        className="odc-pipeline__stage-column"
      >
        <li
          className="odc-pipeline__task"
        >
          <div
            className="title"
          >
            <span
              className="truncate"
            >
              build-image-1
            </span>
            <div
              className="pf-c-tooltip"
              role="tooltip"
            >
              <div
                className="pf-c-tooltip__arrow"
              />
              <div
                className="pf-c-tooltip__content"
              />
            </div>
          </div>
        </li>
        <li
          className="odc-pipeline__task"
        >
          <div
            className="title"
          >
            <span
              className="truncate"
            >
              build-image-2
            </span>
            <div
              className="pf-c-tooltip"
              role="tooltip"
            >
              <div
                className="pf-c-tooltip__arrow"
              />
              <div
                className="pf-c-tooltip__content"
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="odc-pipeline__stage"
    >
      <ul
        className="odc-pipeline__stage-column"
      >
        <li
          className="odc-pipeline__task"
        >
          <div
            className="title"
          >
            <span
              className="truncate"
            >
              openshift-cli-deploy-mapit
            </span>
            <div
              className="pf-c-tooltip"
              role="tooltip"
            >
              <div
                className="pf-c-tooltip__arrow"
              />
              <div
                className="pf-c-tooltip__content"
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
